metricbeat/module/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,108 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+
+	p "github.com/elastic/beats/v7/metricbeat/helper/prometheus"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+func init() {
+	// Register the ModuleFactory function for the "kubernetes" module.
+	if err := mb.Registry.AddModule("kubernetes", ModuleBuilder()); err != nil {
+		panic(err)
+	}
+}
+
+type Module interface {
+	mb.Module
+	GetSharedFamilies(prometheus p.Prometheus) ([]*dto.MetricFamily, error)
+}
+
+type familiesCache struct {
+	sharedFamilies     []*dto.MetricFamily
+	lastFetchErr       error
+	lastFetchTimestamp time.Time
+}
+
+type kubeStateMetricsCache struct {
+	cacheMap map[string]*familiesCache
+	lock     sync.Mutex
+}
+
+func (c *kubeStateMetricsCache) initCacheMapEntry(hash string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, ok := c.cacheMap[hash]; !ok {
+		c.cacheMap[hash] = &familiesCache{}
+	}
+}
+
+type module struct {
+	mb.BaseModule
+
+	kubeStateMetricsCache *kubeStateMetricsCache
+}
+
+func ModuleBuilder() func(base mb.BaseModule) (mb.Module, error) {
+	kubeStateMetricsCache := &kubeStateMetricsCache{
+		cacheMap: make(map[string]*familiesCache),
+	}
+	return func(base mb.BaseModule) (mb.Module, error) {
+		hash := generateCacheHash(base.Config().Hosts)
```

Comment:
In principle the hash is always going to be the same during the life of this module. Wdyt about storing it in `module{}` so it doesn't need to be recalculated every time? Actually, for the same reason, the module could keep a reference to the cache entry directly.